File: python/run-length-encoding/run_length_encoding.py
def encode(string):
    result = ''
    prev, count = None, -1
    for i in range(len(string) + 1):
        if i < len(string) and string[i] == prev:
            count += 1
        else:
            if count > 0:
                if count != 1:
                    result += str(count)
                result += prev
            
            if i < len(string):
                prev, count = string[i], 1
    return result

def decode(string):
    result = ''
    index = 0
    while index < len(string):
        end_index = index
        while end_index < len(string) and string[end_index].isdigit():
            end_index += 1
            
        count = None
        if end_index == index:
            count = 1
        else:
            count = int(string[index:end_index])
        
        for i in range(count):
            result += string[end_index]
        
        index = end_index + 1
    return result

File: java/run-length-encoding/src/main/java/RunLengthEncoding.java
public class RunLengthEncoding {
  String encode(String s) {
    StringBuilder result = new StringBuilder();
    char current = 0;
    int count = 0;
    for (int i = 0; i <= s.length(); ++i) {
      if (i != s.length() && s.charAt(i) == current) {
        ++count;
      } else {
        if (count != 0) {
          if (count != 1) {
            result.append(count);
          }
          result.append(current);
        }

        if (i != s.length()) {
          current = s.charAt(i);
          count = 1;
        }
      }
    }

    return result.toString();
  }

  String decode(String s) {
    StringBuilder result = new StringBuilder();
    int index = 0;
    while (index != s.length()) {
      int endIndex = index;
      while (endIndex != s.length() && Character.isDigit(s.charAt(endIndex))) {
        ++endIndex;
      }

      int count = (endIndex == index) ? 1 : Integer.parseInt(s.substring(index, endIndex));
      result.append(String.valueOf(s.charAt(endIndex)).repeat(count));

      index = endIndex + 1;
    }

    return result.toString();
  }
}
